File: 백준/Gold/12837. 가계부 （Hard）/가계부 （Hard）.cc
// Authored by : chjh2129
#include <bits/stdc++.h>

using namespace std;

using ll = long long;

int n, m;
ll tree[2'000'005];

void update(int p, ll val){
  p += n;
  tree[p] += val;
  for(int i = p; i > 1; i >>= 1) tree[i >> 1] = tree[i] + tree[i ^ 1];
}

ll sum_query(int l, int r){
  ll ret{};
  for(l += n, r += n; l < r; l >>= 1, r >>= 1){
    if(l & 1) ret += tree[l++];
    if(r & 1) ret += tree[--r];
  }
  return ret;
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m;

  // Query
  while(m--){
    ll cmd, a, b;
    cin >> cmd >> a >> b;

    if(cmd&1) update(a-1, b);
    else cout << sum_query(a-1, b) << '\n';
  }
}

File: 백준/Gold/12837. 가계부 （Hard）/가계부 （Hard）.py
#Authored by : chjh2129
import sys
'''
  갱신 함수 update를 값 변경이 아닌 덧셈 연산으로 변경하여 풀이
'''

input = sys.stdin.readline

def update(p, val, tree, n) -> None:
  p += n
  tree[p] += val # 변경이 아닌, 덧셈 연산 수행
  while p > 1:
    tree[p >> 1] = tree[p] + tree[p ^ 1]
    p >>= 1

def sumQuery(l, r, tree, n) -> int:
  ret = 0
  l, r = l + n, r + n
  while l < r:
    if l & 1:
      ret += tree[l]
      l += 1
    if r & 1:
      r -= 1
      ret += tree[r]
    l >>= 1
    r >>= 1
  return ret

def main() -> None:
  # input & init
  n, m = map(int, input().split())

  tree = [ 0 for _ in range(2 * n) ]

  # Query
  for cmd, a, b in (map(int, input().split()) for _ in range(m)):
    if cmd & 1: update(a-1, b, tree, n)
    else:
      print(sumQuery(a-1, b, tree, n))
  
sys.exit(main())